File: scripts_py/send_newsletters.py
"""Rank events for each user and send personalized newsletters via Gmail SMTP."""

from python_app.email_sender import EmailDeliveryService
from python_app.recommender import TwoTowerRecommender
from python_app.supabase_client import ensure_ok, supabase


def main() -> None:
  print("Starting newsletter generation and delivery...")
  try:
    users_resp = supabase.table("users").select("*").execute()
    ensure_ok(users_resp, action="select users")
    users = users_resp.data or []
    print(f"Found {len(users)} users")

    # Use Two-Tower recommender for semantic embedding-based recommendations
    recommender = TwoTowerRecommender()
    print("Loading event embeddings into vector index...")
    event_count = recommender.load_event_index()
    print(f"Loaded {event_count} events into recommendation engine")

    email_service = EmailDeliveryService()
    ranked_events_by_user: dict[str, list[dict]] = {}

    for user in users:
      try:
        recommendations = recommender.get_recommendations(user["id"])
        if recommendations:
          ranked_events_by_user[user["id"]] = recommendations
          print(f"  Generated {len(recommendations)} recommendations for user {user.get('email', user['id'])}")
      except Exception as exc:
        print(f"Failed to get recommendations for user {user.get('id')}: {exc}")

    print(f"Generated recommendations for {len(ranked_events_by_user)} users")
    
    if ranked_events_by_user:
      email_service.send_newsletters(ranked_events_by_user)
    else:
      print("No recommendations generated - skipping email delivery")
    
    print("Newsletter delivery complete!")
  except Exception as exc:
    print(f"Error during newsletter sending: {exc}")
    raise SystemExit(1)


if __name__ == "__main__":
  main()

File: scripts/send-newsletters.ts
import 'dotenv/config';
import { supabase } from '@/lib/supabase';
import { EventScoringService } from '@/lib/scoring';
import { EmailDeliveryService } from '@/lib/email-delivery';

async function sendNewsletters() {
  console.log('Starting newsletter generation and delivery...');

  try {
    // Fetch all users
    const { data: users, error: usersError } = await supabase
      .from('users')
      .select('*');

    if (usersError || !users) {
      throw new Error(`Failed to fetch users: ${usersError?.message}`);
    }

    console.log(`Found ${users.length} users`);

    // Initialize services
    const scoringService = new EventScoringService();
    const emailService = new EmailDeliveryService();

    // Rank events for each user
    const rankedEventsByUser = new Map();

    for (const user of users) {
      try {
        const rankedEvents = await scoringService.rankEventsForUser(user.id, 10);
        if (rankedEvents.length > 0) {
          rankedEventsByUser.set(user.id, rankedEvents);
        }
      } catch (error: any) {
        console.error(`Failed to rank events for user ${user.id}:`, error.message);
      }
    }

    console.log(`Ranked events for ${rankedEventsByUser.size} users`);

    // Send emails
    await emailService.sendNewsletters(rankedEventsByUser);

    console.log('✓ Newsletter delivery complete!');
  } catch (error: any) {
    console.error('Error during newsletter sending:', error.message);
    process.exit(1);
  }
}

sendNewsletters();
